src/crowdin_client.cpp

```diff
@@ -135,150 +159,332 @@ CrowdinClient::~CrowdinClient() {}
+                i.pathName = str::to_wstring(pathStr) + i.pathName;
+            }
+        }
+
+        //TODO: get more until all branches gotten (if more than 500)    
+        return m_api->get(url + "/branches?limit=500");
+    }).then([this, url, prj](json r) {
+        // Handle branches
+        map<int, string> branches;
+
+        for(const auto& i : r["data"]) {
+            const json& d = i["data"];
+            branches[d["id"]] = d["name"].get<string>();
+        }
+
+        for(auto& i : prj->files) {
+            if(i.branchId != NO_ID) {
+                i.pathName = "/" + str::to_wstring(branches[i.branchId]) + i.pathName;
+            }
+        }
+
+        return *prj;
+    });
 }
 
 
-dispatch::future<void> CrowdinClient::DownloadFile(const std::string& project_id,
-                                                   const std::wstring& file,
-                                                   const Language& lang,
+dispatch::future<void> CrowdinClient::DownloadFile(const long project_id,
+                                                   const long file_id,
+                                                   const wxString& file_name,
+                                                   const std::string& lang_tag,
                                                    const std::wstring& output_file)
 {
-    // NB: "export_translated_only" means the translation is not filled with the source string
-    //     if there's no translation, i.e. what Poedit wants.
-    auto url = "/api/project/" + project_id + "/export-file"
-                   "?json="
-                   "&export_translated_only=1"
-                   "&language=" + lang.LanguageTag() +
-                   "&file=" + http_client::url_encode(file);
-    return m_api->download(url, output_file);
+    cout << "\n\nGetting file URL: " << "/projects/" + std::to_string(project_id) + "/files/" + std::to_string(file_id) + "/download" << "\n\n";
+    return RefreshToken().then([=] () {
+    return m_api->post(
+        "projects/" + std::to_string(project_id) + "/translations/builds/files/" + std::to_string(file_id),
+        json_data(json({
+            { "targetLanguageId", lang_tag },
+            // for XLIFF files should be exported "as is" so set to `false`
+            { "exportAsXliff", !file_name.Lower().EndsWith(".xliff.xliff") }
+        })))
+        .then([this, output_file] (json r) {
+            string url = r["data"]["url"],
+                   proto(wxString(url).BeforeFirst(':').mb_str()),
+                   host(wxString(url).AfterFirst('/').AfterFirst('/').BeforeFirst('/').mb_str());
+
+            auto downloader = make_shared<crowdin_http_client>(*this, proto + "://" + host);
+            cout << "\n\nGotten file URL: "<< r << "\n\n";
+            return downloader->download(url, output_file)
+                .then([downloader]() {
+                    return dispatch::make_ready_future();
+                });
+        });
```

Comment:
I think this creates memory leak — the future has a `shared_ptr` to the downloader, and is called by it, so refcount is always >= 1 and neither the future nor the client ever get released.

Reply:
Not sure why since below output shows opposite (just providing as is without cutting anything)
```
Getting file URL: /projects/4/files/284/download
http_client CONSTRUCTED
TestMemoryLeaks CONSTRUCTED
Gotten file URL: {"data":{"etag":null,"expireIn":"2020-04-26T09:12:02+00:00","url":"https://production-enterprise-tmp.downloads.crowdin.com/exported_files/strings.xliff?response-content-disposition=attachment%3B%20filename%3D%22strings.xliff%22&X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIGJKLQV66ZXPMMEA%2F20200426%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20200426T091002Z&X-Amz-SignedHeaders=host&X-Amz-Expires=120&X-Amz-Signature=284f58e448ceeed0a5fa605d083820080bd8252054d2a48a42fb044048ddfc3b"}}
Dowloading file: https://production-enterprise-tmp.downloads.crowdin.com/exported_files/strings.xliff?response-content-disposition=attachment%3B%20filename%3D%22strings.xliff%22&X-Amz-Content-Sha256=UNSIGNED-PAYLOAD&X-Amz-Algorithm=AWS4-HMAC-SHA256&X-Amz-Credential=AKIAIGJKLQV66ZXPMMEA%2F20200426%2Fus-east-1%2Fs3%2Faws4_request&X-Amz-Date=20200426T091002Z&X-Amz-SignedHeaders=host&X-Amz-Expires=120&X-Amz-Signature=284f58e448ceeed0a5fa605d083820080bd8252054d2a48a42fb044048ddfc3b
Downloading file response: 200, application/octet-stream
Downloading file: closing
TestMemoryLeaks DESTRUCTED
http_client DESTRUCTED

```
Produced by executable build from this test branch https://github.com/tetatetit/poedit/tree/test-downloader-memory-leak on Linux

Reply:
Marking it as resolved unless you still see anything to discuss/improve/correct/fix here

Reply:
Yeah, sorry, that was a brainfart on my part :( 

It does open another question, though: why do you keep around `m_oauth` HTTP client, instead of using a transient one like this too? Because it is expected to be used more often (which with implicit flow probably won't be the case any longer?)?

Reply:
yep, exactly, since it needed to `refresh token` every few hours before any API call. Though now when implicit flow getting back - it's really needless as class member but just local (func) scope.